cicecore/cicedynB/dynamics/ice_dyn_shared.F90

```diff
@@ -1358,144 +1905,1025 @@ subroutine strain_rates (nx_block,   ny_block,   &
       
-      real (kind=dbl_kind), intent(out):: &  
-        zetax2ne, zetax2nw, zetax2sw, zetax2se,  & ! zetax2 at each corner 
-        etax2ne, etax2nw, etax2sw, etax2se,      & ! etax2 at each corner
-        rep_prsne, rep_prsnw, rep_prssw, rep_prsse ! replacement pressure
+      subroutine visc_replpress_avgzeta (zetax2T1, zetax2T2, &
+                                         zetax2T3, zetax2T4, &
+                                          etax2T1,  etax2T2, &
+                                          etax2T3,  etax2T4, & 
+                                            mask1,    mask2, &
+                                            mask3,    mask4, &
+                                            area1,    area2, &
+                                            area3,    area4, &
+                            deltaU, zetax2U, etax2U, rep_prsU)
+
+      real (kind=dbl_kind), intent(in):: &
+         zetax2T1,zetax2T2,zetax2T3,zetax2T4, &
+          etax2T1, etax2T2, etax2T3, etax2T4, &
+            mask1,   mask2,   mask3,   mask4, &
+            area1,   area2,   area3,   area4, &
+         deltaU
+
+      real (kind=dbl_kind), optional, intent(out):: &
+         zetax2U, etax2U, rep_prsU ! 2 x viscosities, replacement pressure
 
       ! local variables
+
       real (kind=dbl_kind) :: &
-        tmpcalcne, tmpcalcnw, tmpcalcsw, tmpcalcse
+         lzetax2U, &  ! local variable
+         areatmp
+
+      character(len=*), parameter :: subname = '(visc_replpress_avgzeta)'
 
       ! NOTE: for comp. efficiency 2 x zeta and 2 x eta are used in the code
 
-      ! if (trim(yield_curve) == 'ellipse') then       
-        tmpcalcne = capping     *(strength/max(Deltane, tinyarea))+ &
-                    (c1-capping)* strength/   (Deltane+ tinyarea)   
-        tmpcalcnw = capping     *(strength/max(Deltanw, tinyarea))+ &
-                    (c1-capping)* strength/   (Deltanw+ tinyarea)   
-        tmpcalcsw = capping     *(strength/max(Deltasw, tinyarea))+ &
-                    (c1-capping)* strength/   (Deltasw+ tinyarea)  
-        tmpcalcse = capping     *(strength/max(Deltase, tinyarea))+ &
-                    (c1-capping)* strength/   (Deltase+ tinyarea)
-
-        zetax2ne  = (c1+Ktens)*tmpcalcne ! northeast 
-        rep_prsne = (c1-Ktens)*tmpcalcne*Deltane
-        etax2ne   = epp2i*zetax2ne
-         
-        zetax2nw  = (c1+Ktens)*tmpcalcnw ! northwest 
-        rep_prsnw = (c1-Ktens)*tmpcalcnw*Deltanw
-        etax2nw   = epp2i*zetax2nw
+      areatmp = (mask1 * area1   + &
+                 mask4 * area4   + &
+                 mask3 * area3   + &
+                 mask2 * area2)
+
+      if (present(rep_prsU) .or. present(zetax2U)) then
+         lzetax2U = (mask1 * area1 * zetax2T1  + &
+                     mask4 * area4 * zetax2T4  + &
+                     mask3 * area3 * zetax2T3  + &
+                     mask2 * area2 * zetax2T2) / areatmp
+         if (present(zetax2U)) then
```

Comment:
As noted in ice_dyn_evp.F90 comments, these special visc_repl* routines seem very expensive -- called from within subroutine loops for what appears to be array-level calculations, lots of if statements.  Maybe you can use existing averaging routines and inline the differences, for better efficiency?

Reply:
This has been refactored leverage grid_average_X2Y and removing optional arguments.